lib/smooth/core/io/i2c/I2CMasterDevice.cpp

```diff
@@ -24,77 +24,148 @@ limitations under the License.
 #include "esp_intr_alloc.h"
 
 using namespace smooth::core::logging;
+using namespace smooth::core::util;
 
 namespace smooth::core::io::i2c
 {
     // Understanding I2C: http://www.ti.com/lit/an/slva704/slva704.pdf
     // I2C specification: http://www.nxp.com/docs/en/user-guide/UM10204.pdf
+    // Clock Stretching:  https://www.i2c-bus.org/clock-stretching/
 
     constexpr const char* log_tag = "I2CMasterDevice";
+    constexpr const char* write_tag = "write";
+    constexpr const char* read_tag = "read";
+    constexpr const char* read8_tag = "read8";
+    constexpr const char* read16_tag = "read16";
+    constexpr const char* read_block_tag = "read_block";
     constexpr const std::chrono::milliseconds timeout(1000);
 
+    // Write address and data
     bool I2CMasterDevice::write(uint8_t address, std::vector<uint8_t>& data, bool expect_ack)
     {
         I2CCommandLink link(*this);
 
+        set_process_name(write_tag);
+
         // Set R/W bit to 0 for write.
-        address = static_cast<uint8_t>(address << 1);
+        auto write_address = static_cast<uint8_t>(address << 1);
 
         auto res = i2c_master_start(link);
-        res |= i2c_master_write_byte(link, address, expect_ack);
+        res |= i2c_master_write_byte(link, write_address, expect_ack);
         res |= i2c_master_write(link, data.data(), data.size(), expect_ack);
         res |= i2c_master_stop(link);
+        res |= i2c_master_cmd_begin(port, link, to_tick(timeout));
```

Comment:
Funny how I've not seen this before. These should have been `&=`, but with the new learning as noted in another comment, we need to use &&.

```suggestion
auto res = i2c_master_start(link)
        && i2c_master_write_byte(link, write_address, expect_ack);
        && i2c_master_write(link, data.data(), data.size(), expect_ack);
        && i2c_master_stop(link);
        && i2c_master_cmd_begin(port, link, to_tick(timeout));
```